addOns/dns/src/main/java/org/zaproxy/addon/dns/SpfScanner.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.dns;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.commons.httpclient.URIException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractHostPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.addon.dns.exceptions.TooManyRecords;
+
+public class SpfScanner extends AbstractHostPlugin {
+    private static final int id = 90040; // assigned on ZAPROXY repository
+    private static final Logger LOGGER = LogManager.getLogger(SpfScanner.class);
+    private static final String MESSAGE_PREFIX = "dns.";
+    private volatile boolean enabled = true;
+    private static List<String> reviewedDomains = new ArrayList<String>();
+
+    private String getConstantString(String key) {
+        return Constant.messages.getString(MESSAGE_PREFIX + key);
+    }
+
+    private String getHigherSubdomain(String host) {
+        String[] hostarray = host.split("\\.");
+        if (hostarray.length < 2) {
+            return null;
+        }
+        return String.join(".", Arrays.copyOfRange(hostarray, 1, hostarray.length));
+    }
```

Comment:
I wonder if Java or some library has good handling for this. Mainly I'm thinking of cases like `mystore.co.uk` vs `www.mystore.co.uk`, etc.

Reply:
SPF is hierarchical. So here I'm reviewing it hierarchically.

This method just removes a subdomain, and the algorithm at `findValidSpfRecord` will call it until it finds a SPF record.

For instance... it might call for `www.mystore.co.uk` and find it, and later search for `mystore.co.uk`, what might not have it, falling back to `co.uk` , what is fine. The algorithm should work fine in both cases. Anyway... I plan to perform a refactor and move both methods to `SPFParser` because it is going to require a `DNSClient` anyway.

Reply:
Maybe I'm over thinking it. Is there stuff hosted directly at .uk or are .co.uk and .com functionally equivalent. (mystore.com vs mystore.co.uk or vs just .co.uk).

Also if you return "" vs null then you're less likely to hit an NPE. (Granted isempty and !=null are hardly different but still ... I guess you get to avoid negative logic).

Reply:
AFAIK, `.co.uk` and `.com` are not different domains on different providers.

For `.co.uk`, the algorithm would ask even to `.uk`, what requires an extra DNS request but will always return an empty list, so it should work any way.